Sources/Tracking/Background Queue/QueueQueryRunner.swift

```diff
@@ -0,0 +1,50 @@
+import Foundation
+
+internal protocol QueueQueryRunner: AutoMockable {
+    func getNextTask(_ queue: [QueueTaskMetadata], lastFailedTask: QueueTaskMetadata?) -> QueueTaskMetadata?
+}
+
+// sourcery: InjectRegister = "QueueQueryRunner"
+internal class CioQueueQueryRunner: QueueQueryRunner {
+    func getNextTask(_ queue: [QueueTaskMetadata], lastFailedTask: QueueTaskMetadata?) -> QueueTaskMetadata? {
+        guard !queue.isEmpty else {
+            return nil
+        }
+
+        guard let lastFailedTask = lastFailedTask else {
+            return queue[0]
+        }
+
+        let criteria = getCriteria(lastFailedTask: lastFailedTask)
+
+        return queue.first(where: { doesTaskPassCriteria($0, criteria: criteria) })
+    }
+
+    private func getCriteria(lastFailedTask: QueueTaskMetadata) -> QueueQueryCriteria {
+        var criteria = QueueQueryCriteria(excludeGroups: nil)
+
+        if let groupsToExclude = lastFailedTask.groupsParent {
```

Comment:
since this is only considering the last failed task, my reading of it is that if you had (using my suggested terminology - `taskName` below is equiv of `groupsParent` and `blockingTaskNames` is `groupsChild`

```
t1,taskName:X,failed
t2,blockingTaskNames:[X],skipped
t3,taskName:Y,failed
t4,blockingTaskNames:[X]
```

then we would execute `t4` because I don't see anything in here accumulating a list of `taskNames` that are now blocked

Reply:
Good catch. This is now fixed where we keep track of *all* failed group names not just the last one.